Datastore/src/Operation.php

```diff
@@ -491,26 +566,23 @@ public function runQuery(QueryInterface $query, array $options = [])
         }
         $runQueryObj = clone $query;
         $runQueryFn = function (array $args = []) use (&$runQueryObj, $options, &$remainingLimit) {
-            $args += [
-                'query' => [],
-            ];
+            $parsedArgs = $this->parseRunQueryArguments($args, $runQueryObj, $options, $remainingLimit);
 
-            // The iterator provides the startCursor for subsequent pages as an argument.
-            $requestQueryArr = $args['query'] + $runQueryObj->queryObject();
-            if (isset($remainingLimit)) {
-                $requestQueryArr['limit'] = $remainingLimit;
-            }
-            $request = [
-                'projectId' => $this->projectId,
-                'partitionId' => $this->partitionId(
-                    $this->projectId,
-                    $options['namespaceId'],
-                    $options['databaseId']
-                ),
-                $runQueryObj->queryKey() => $requestQueryArr,
-            ] + $this->readOptions($options) + $options;
-
-            $res = $this->connection->runQuery($request);
+            list($data, $optionalArgs) = $this->splitOptionalArgs($parsedArgs);
+
+            // Remove redundant keys for request.
+            $this->pluckArray(
+                ['className', 'namespaceId', 'readTime', 'readConsistency', 'transaction'],
+                $data
+            );
```

Comment:
I don't understand what's happening here - why do we remove these fields from the request? What makes them redudant?

Reply:
There were many field which are a nested level in the Request. For example, `readTime`, `readConsistency` and `transaction` are a part of `readOptions`. We exposed these options directly to the user and internally made sure that it's properly populated in the readOptions.

Similar is the case with `namespaceId`, it's part of `partitionId` field which is the actual part of the request.

Reply:
I see, the function `readOptions` essentially moves these fields to the `readOptions` array (I suppose as a "convenience"), and so we have to _remove_ those fields explicitly otherwise some error is thrown?